### PR TITLE
Fix win amd64 dtypes

### LINKED ISSUE #36: Many test errors of scikits.umfpack 0.3.0 on win-amd64
The 32-bit builds are OK. The 64-bit builds fail most tests:
```
Running unit tests for scikits.umfpack
NumPy version 1.13.1
NumPy relaxed strides checking option: True
NumPy is installed in X:\Python36\lib\site-packages\numpy
Python version 3.6.1 (v3.6.1:69c0db5, Mar 21 2017, 18:41:36) [MSC v.1900 64 bit (AMD64)]
nose version 1.3.7
EEEEE.EEEX:\Python36\lib\importlib\_bootstrap.py:205: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
E.E.EE.EEEEE
======================================================================
ERROR: test_interface.TestSolvers.test_solve_complex_int64_umfpack
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\numpy\testing\decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_interface.py", line 56, in test_solve_complex_int64_umfpack
    x = um.spsolve(a, b)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 56, in spsolve
    x = UmfpackLU(A).solve(b)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 192, in __init__
    self.umf = UmfpackContext(family[A.dtype.char + A.indices.dtype.char])
KeyError: 'Dq'

======================================================================
ERROR: test_interface.TestSolvers.test_solve_complex_umfpack
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_interface.py", line 48, in test_solve_complex_umfpack
    x = um.spsolve(a, b)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 56, in spsolve
    x = UmfpackLU(A).solve(b)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 193, in __init__
    self.umf.numeric(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 539, in numeric
    self.symbolic(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 520, in symbolic
    umfStatus[status]))
RuntimeError: <function umfpack_zl_symbolic at 0x00000244E656E0D0> failed with UMFPACK_ERROR_out_of_memory

======================================================================
ERROR: test_interface.TestSolvers.test_solve_int64_umfpack
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\numpy\testing\decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_interface.py", line 72, in test_solve_int64_umfpack
    x = um.spsolve(a, b)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 56, in spsolve
    x = UmfpackLU(A).solve(b)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 192, in __init__
    self.umf = UmfpackContext(family[A.dtype.char + A.indices.dtype.char])
KeyError: 'dq'

======================================================================
ERROR: test_interface.TestSolvers.test_solve_sparse_rhs
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_interface.py", line 79, in test_solve_sparse_rhs
    x = um.spsolve(a, b)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 56, in spsolve
    x = UmfpackLU(A).solve(b)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 193, in __init__
    self.umf.numeric(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 539, in numeric
    self.symbolic(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 520, in symbolic
    umfStatus[status]))
RuntimeError: <function umfpack_dl_symbolic at 0x00000244E656DF28> failed with UMFPACK_ERROR_invalid_matrix

======================================================================
ERROR: test_interface.TestSolvers.test_solve_umfpack
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_interface.py", line 63, in test_solve_umfpack
    x = um.spsolve(a, b)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 56, in spsolve
    x = UmfpackLU(A).solve(b)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 193, in __init__
    self.umf.numeric(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 539, in numeric
    self.symbolic(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 520, in symbolic
    umfStatus[status]))
RuntimeError: <function umfpack_dl_symbolic at 0x00000244E656DF28> failed with UMFPACK_ERROR_out_of_memory

======================================================================
ERROR: test_interface.TestSolvers.test_splu_solve
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_interface.py", line 85, in test_splu_solve
    lu = um.splu(a)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 84, in splu
    return UmfpackLU(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 193, in __init__
    self.umf.numeric(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 539, in numeric
    self.symbolic(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 520, in symbolic
    umfStatus[status]))
RuntimeError: <function umfpack_dl_symbolic at 0x00000244E656DF28> failed with UMFPACK_ERROR_out_of_memory

======================================================================
ERROR: test_interface.TestSolvers.test_splu_solve_int64
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\numpy\testing\decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_interface.py", line 97, in test_splu_solve_int64
    lu = um.splu(a)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 84, in splu
    return UmfpackLU(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 192, in __init__
    self.umf = UmfpackContext(family[A.dtype.char + A.indices.dtype.char])
KeyError: 'dq'

======================================================================
ERROR: test_interface.TestSolvers.test_splu_solve_sparse
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_interface.py", line 107, in test_splu_solve_sparse
    lu = um.splu(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 84, in splu
    return UmfpackLU(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\interface.py", line 193, in __init__
    self.umf.numeric(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 539, in numeric
    self.symbolic(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 520, in symbolic
    umfStatus[status]))
RuntimeError: <function umfpack_dl_symbolic at 0x00000244E656DF28> failed with UMFPACK_ERROR_invalid_matrix

======================================================================
ERROR: test_umfpack.TestFactorization.test_complex_int64_lu
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\numpy\testing\decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_umfpack.py", line 160, in test_complex_int64_lu
    umfpack.numeric(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 539, in numeric
    self.symbolic(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 499, in symbolic
    indx = self._getIndx(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 478, in _getIndx
    raise ValueError('matrix must have long indices')
ValueError: matrix must have long indices

======================================================================
ERROR: test_umfpack.TestFactorization.test_real_int64_lu
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\numpy\testing\decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_umfpack.py", line 201, in test_real_int64_lu
    umfpack.numeric(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 539, in numeric
    self.symbolic(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 499, in symbolic
    indx = self._getIndx(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 478, in _getIndx
    raise ValueError('matrix must have long indices')
ValueError: matrix must have long indices

======================================================================
ERROR: test_umfpack.TestScipySolvers.test_factorized_long_umfpack
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\numpy\testing\decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_umfpack.py", line 104, in test_factorized_long_umfpack
    solve = linsolve.factorized(a)
  File "X:\Python36\lib\site-packages\scipy\sparse\linalg\dsolve\linsolve.py", line 376, in factorized
    umf = umfpack.UmfpackContext(_get_umf_family(A))
  File "X:\Python36\lib\site-packages\scipy\sparse\linalg\dsolve\linsolve.py", line 60, in _get_umf_family
    return family[dt]
KeyError: 'dq'

======================================================================
ERROR: test_umfpack.TestScipySolvers.test_factorized_umfpack
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_umfpack.py", line 92, in test_factorized_umfpack
    solve = linsolve.factorized(a)
  File "X:\Python36\lib\site-packages\scipy\sparse\linalg\dsolve\linsolve.py", line 379, in factorized
    umf.numeric(A)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 539, in numeric
    self.symbolic(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 520, in symbolic
    umfStatus[status]))
RuntimeError: <function umfpack_dl_symbolic at 0x00000244E656DF28> failed with UMFPACK_ERROR_invalid_matrix

======================================================================
ERROR: test_umfpack.TestScipySolvers.test_solve_complex_long_umfpack
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\numpy\testing\decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_umfpack.py", line 60, in test_solve_complex_long_umfpack
    x = linsolve.spsolve(a, b)
  File "X:\Python36\lib\site-packages\scipy\sparse\linalg\dsolve\linsolve.py", line 143, in spsolve
    umf = umfpack.UmfpackContext(_get_umf_family(A))
  File "X:\Python36\lib\site-packages\scipy\sparse\linalg\dsolve\linsolve.py", line 60, in _get_umf_family
    return family[dt]
KeyError: 'Dq'

======================================================================
ERROR: test_umfpack.TestScipySolvers.test_solve_complex_umfpack
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_umfpack.py", line 51, in test_solve_complex_umfpack
    x = linsolve.spsolve(a, b)
  File "X:\Python36\lib\site-packages\scipy\sparse\linalg\dsolve\linsolve.py", line 145, in spsolve
    autoTranspose=True)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 753, in linsolve
    self.numeric(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 539, in numeric
    self.symbolic(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 520, in symbolic
    umfStatus[status]))
RuntimeError: <function umfpack_zl_symbolic at 0x00000244E656E0D0> failed with UMFPACK_ERROR_invalid_matrix

======================================================================
ERROR: test_umfpack.TestScipySolvers.test_solve_long_umfpack
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\numpy\testing\decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_umfpack.py", line 77, in test_solve_long_umfpack
    x = linsolve.spsolve(a, b)
  File "X:\Python36\lib\site-packages\scipy\sparse\linalg\dsolve\linsolve.py", line 143, in spsolve
    umf = umfpack.UmfpackContext(_get_umf_family(A))
  File "X:\Python36\lib\site-packages\scipy\sparse\linalg\dsolve\linsolve.py", line 60, in _get_umf_family
    return family[dt]
KeyError: 'dq'

======================================================================
ERROR: test_umfpack.TestScipySolvers.test_solve_sparse_rhs
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_umfpack.py", line 85, in test_solve_sparse_rhs
    x = linsolve.spsolve(a, b)
  File "X:\Python36\lib\site-packages\scipy\sparse\linalg\dsolve\linsolve.py", line 145, in spsolve
    autoTranspose=True)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 753, in linsolve
    self.numeric(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 539, in numeric
    self.symbolic(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 520, in symbolic
    umfStatus[status]))
RuntimeError: <function umfpack_dl_symbolic at 0x00000244E656DF28> failed with UMFPACK_ERROR_invalid_matrix

======================================================================
ERROR: test_umfpack.TestScipySolvers.test_solve_umfpack
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python36\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "X:\Python36\lib\site-packages\scikits\umfpack\tests\test_umfpack.py", line 68, in test_solve_umfpack
    x = linsolve.spsolve(a, b)
  File "X:\Python36\lib\site-packages\scipy\sparse\linalg\dsolve\linsolve.py", line 145, in spsolve
    autoTranspose=True)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 753, in linsolve
    self.numeric(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 539, in numeric
    self.symbolic(mtx)
  File "X:\Python36\lib\site-packages\scikits\umfpack\umfpack.py", line 520, in symbolic
    umfStatus[status]))
RuntimeError: <function umfpack_dl_symbolic at 0x00000244E656DF28> failed with UMFPACK_ERROR_invalid_matrix

----------------------------------------------------------------------
Ran 21 tests in 0.141s

FAILED (errors=17)
```

### PR DESCRIPTION
**Pull request by xoviat**
Building on top of the work by @rc, we fix the windows compilation issues through using the C99 `int64_t` type that is supported in both windows and linux. Specifically, we apply the typemaps assigned to long to our new type with `apply long`:

```
+%apply long *OUTPUT {
```

We also tell swig how to convert Python integers to our new type:

```
%typemap(out) SuiteSparse_long {
  $result = PyInt_FromLong((int)$1);
}
```

The result is that we can compile both on windows and linux.

Closes #37.
Closes #36.

### CODE REVIEW

**Comment by xoviat** [comment]
This [did not actually fix the problem](https://ci.appveyor.com/project/rc/scikit-umfpack-wheels/build/1.0.17/job/vsq9t7jk43vtjmj8) but rather introduced new errors. @rc Is there a way to get the UMFpack output? I looked at the source and it should print a message if it fails.

**Comment by rc** [comment]
Yes, the complaint about unsafe cast to int32 does not meant that the data should be int32 in all cases - it has to be int64 for `*int64*` tests. The problem was (and still is) how to define equivalence of int64 and SuiteSparse_long in `umfpack.i` on Windows. On linux, `typedef long SuiteSparse_long;` works. On windows, long is just int32 - I worked around the unsafe cast error by using (unportable)  `typedef __int64 SuiteSparse_long;` on Windows, but then got another error concerning the matrix shape arguments - apparently, swig cannot by itself cast between a Python integer and __int64.

So the correct way to fix this would be to define SuiteSparse_long to be the same as int64 in a portable way. Any suggestions welcome!

**Comment by xoviat** [comment]
I would recommend using `typedef int64_t SuiteSparse_long` with the [stdint header](https://github.com/mattn/gntp-send/blob/090b1276fb4a30601113551bed88c58329646818/include/msinttypes/stdint.h). That's what SciPy uses and it works pretty well.

**Comment by xoviat** [comment]
Let me try that...

**Comment by xoviat** [comment]
@rc This fixes the problems except for two (new) test failures that occur on both Windows and Linux. Can you explain these new failures?

File: scikits/umfpack/umfpack.i [hunk 1]
```diff
@@ -248,6 +265,13 @@ OPAQUE_ARGINOUT( void * )
...
     long *nz_udiag
 };
+%apply SuiteSparse_long *OUTPUT {
+    SuiteSparse_long *lnz,
+    SuiteSparse_long *unz,
+    SuiteSparse_long *n_row,
+    SuiteSparse_long *n_col,
+    SuiteSparse_long *nz_udiag
+};
 %include <umfpack_get_lunz.h>
```
Comment (xoviat) [inline scikits/umfpack/umfpack.i]: Actually, the error is:

```
E       TypeError: umfpack_zl_get_lunz() missing 5 required positional arguments: 'unz', 'n_row', 'n_col', 'nz_udiag', and 'Numeric'
```
That might be related to these changes.

**Comment by rc** [comment]
The error
```
TypeError: umfpack_dl_get_lunz() takes exactly 6 arguments (1 given)
```
is caused by the fact, that the `umfpack_dl_get_lunz()` arguments are now no longer recognized as output arguments.

I tried to build on your latest commit and fix the code on linux with int64_t, but failed  - the only way that works for me is to use the original
```
typedef  long SuiteSparse_long;
```
or
```
typedef int64_t SuiteSparse_long;
typedef long int64_t;
```
together with the original typemaps (or int64_t based typemaps, provided `typedef long int64_t` is in effect).

**Comment by xoviat** [comment]
Maybe we can ask for help from the SWIG folks?

**Comment by rc** [comment]
Good idea, especially those with Windows development experience - what int64 windows C type is recognized by swig in a similar way as long on linux. The wrappers are very old, and I no longer remember how to write proper swig (not to mention moder swig) wrappers. :]

**Comment by xoviat** [comment]
> what int64 windows C type is recognized by swig in a similar way as long on linux

The correct type is `int64_t`. The code is wrong (dependent on the gcc compiler implementation) but it was only ever compiled on gcc so no one noticed. That said, C99 is required to write correct code. So it's possible that the authors couldn't have written it correctly (if it was you or someone else).

**Comment by rc** [comment]
Yes, it was me (many years ago), I confess :).

**Comment by xoviat** [comment]
@rc Should be ready to merge.

File: scikits/umfpack/umfpack.py [hunk 2]
```diff
@@ -505,13 +505,16 @@ def symbolic(self, mtx):
...
+                                          mtx.data,
                                           self.control, self.info)
         else:
             real, imag = mtx.data.real.copy(), mtx.data.imag.copy()
             status, self._symbolic\
                     = self.funs.symbolic(mtx.shape[0], mtx.shape[1],
-                                          mtx.indptr, indx,
+                                          mtx.indptr.astype('int32'),
+                                          indx.astype('int32'),
                                           real, imag,
```
Comment (rc) [inline scikits/umfpack/umfpack.py]: Does it work on Windows when the `.astype('int32')` conversion is reverted? It does on Linux.

This retyping effectively negates  PR #23, i.e. to solve problems with matrices such that, during factorization, the int32 indices are not enough to store all matrix nonzeros.

File: scikits/umfpack/umfpack.py [hunk 3]
```diff
@@ -505,13 +505,16 @@ def symbolic(self, mtx):
         if self.isReal:
             status, self._symbolic\
                     = self.funs.symbolic(mtx.shape[0], mtx.shape[1],
-                                          mtx.indptr, indx, mtx.data,
+                                          mtx.indptr.astype('int32'),
+                                          indx.astype('int32'),
```
Comment (rc) [inline scikits/umfpack/umfpack.py]: See below...

**Comment by rc** [comment]
Thanks, great work! Could you try looking at the `.astype('int32')` problem as described in the code comments? So the trick to make swig recognize int64_t was to `%include "stdint.i"`.

**Comment by xoviat** [comment]
> So the trick to make swig recognize int64_t was to %include "stdint.i".

Actually, the "trick" was to `%apply long` to the new type. If I understand correctly, that applies the type conversions that SWIG uses for `long` to `int64_t`. But SWIG is honestly one of the most confusing languages out there.

**Comment by xoviat** [comment]
@rc Seems to work perfectly now.

**Comment by rc** [comment]
Great, thanks!

**Pull request merged**